File: Trinity.Runtime/platform_src/android/app/src/main/java/org/elastos/trinity/runtime/passwordmanager/PasswordDatabaseInfo.java
package org.elastos.trinity.runtime.passwordmanager;

import org.elastos.trinity.runtime.passwordmanager.passwordinfo.PasswordInfo;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Iterator;

/**
 * Database JSON format:
 *
 * {
 *     "applications": {
 *          "APPIID1": {
 *              "passwordentries": [
 *                  {
 *                      RAW_USER_OBJECT
 *                  }
 *              ]
 *          }
 *     }
 * }
 *
 * We work directly with raw JSONObjects to make it easier later to maintain the structure, add new fields,
 * handle specific or missing items.
 */
class PasswordDatabaseInfo {
    private static final String APPLICATIONS_KEY = "applications";
    private static final String PASSWORD_ENTRIES_KEY = "passwordentries";
    JSONObject rawJson;
    String activeMasterPassword = null;

    static PasswordDatabaseInfo createEmpty() {
        try {
            PasswordDatabaseInfo info = new PasswordDatabaseInfo();
            JSONObject applications = new JSONObject();
            info.rawJson = new JSONObject();
            info.rawJson.put(APPLICATIONS_KEY, applications);
            return info;
        }
        catch (Exception e) {
            e.printStackTrace();
            return null;
        }
    }

    public static PasswordDatabaseInfo fromJson(String json) throws JSONException {
        PasswordDatabaseInfo info = new PasswordDatabaseInfo();
        info.rawJson = new JSONObject(json);
        return info;
    }

    public PasswordInfo getPasswordInfo(String appID, String key) throws Exception {
        JSONObject appIDContent = getAppIDContent(appID);
        if (appIDContent == null) {
            // No entry for this app ID yet, so we can't find the requested key
            return null;
        }

        JSONArray passwordEntries = appIDContent.getJSONArray(PASSWORD_ENTRIES_KEY);
        JSONObject entry = passwordEntry(passwordEntries, key);
        if (entry == null) {
            // No such entry exists
            return null;
        }

        return PasswordInfoBuilder.buildFromType(entry);
    }

    public void setPasswordInfo(String appID, PasswordInfo info) throws JSONException {
        JSONObject appIDContent = getAppIDContent(appID);
        if (appIDContent == null) {
            // No entry for this app ID yet, create one and add it
            appIDContent = createdEmptyAppIDContent();
            JSONObject applications = rawJson.getJSONObject(APPLICATIONS_KEY);
            applications.put(appID, appIDContent);
        }

        JSONArray passwordEntries = appIDContent.getJSONArray(PASSWORD_ENTRIES_KEY);
        if (keyInPasswordEntries(passwordEntries, info.key)) {
            // This entry already exists. Delete it first before re-adding its updated version.
            deletePasswordEntryFromKey(passwordEntries, info.key);
        }
        addPasswordEntry(passwordEntries, info);
    }

    public ArrayList<PasswordInfo> getAllPasswordInfo() throws Exception {
        JSONObject applications = rawJson.getJSONObject(APPLICATIONS_KEY);

        ArrayList<PasswordInfo> infos = new ArrayList<>();
        Iterator<String> it = applications.keys();
        while (it.hasNext()) {
            String appID = it.next();
            JSONObject appIDContent = getAppIDContent(appID);
            if (appIDContent != null) {
                JSONArray passwordEntries = appIDContent.getJSONArray(PASSWORD_ENTRIES_KEY);
                for (int i=0; i<passwordEntries.length(); i++) {
                    JSONObject entry = passwordEntries.getJSONObject(i);
                    PasswordInfo info = PasswordInfoBuilder.buildFromType(entry);
                    if (info != null) {
                        infos.add(info);
                    }
                }
            }
        }
        return infos;
    }

    public void deletePasswordInfo(String appID, String key) throws JSONException {
        JSONObject appIDContent = getAppIDContent(appID);
        if (appIDContent == null) {
            // No entry for this app ID yet, so we can't find the requested key
            return;
        }

        JSONArray passwordEntries = appIDContent.getJSONArray(PASSWORD_ENTRIES_KEY);
        deletePasswordEntryFromKey(passwordEntries, key);
    }

    private JSONObject getAppIDContent(String appID) throws JSONException {
        JSONObject applications = rawJson.getJSONObject(APPLICATIONS_KEY);
        if (applications.has(appID)) {
            return applications.getJSONObject(appID);
        }
        else {
            return null;
        }
    }

    private JSONObject createdEmptyAppIDContent() throws JSONException {
        JSONObject appIDContent = new JSONObject();
        appIDContent.put(PASSWORD_ENTRIES_KEY, new JSONArray());
        return appIDContent;
    }

    private JSONObject passwordEntry(JSONArray entries, String key) throws JSONException {
        for (int i=0; i<entries.length(); i++) {
            JSONObject info = entries.getJSONObject(i);
            if (info.getString("key").equals(key))
                return info;
        }
        return null;
    }

    private int passwordEntryIndex(JSONArray entries, String key) throws JSONException {
        for (int i=0; i<entries.length(); i++) {
            JSONObject info = entries.getJSONObject(i);
            if (info.getString("key").equals(key))
                return i;
        }
        return -1;
    }

    private boolean keyInPasswordEntries(JSONArray entries, String key) throws JSONException {
        return passwordEntryIndex(entries, key) >= 0;
    }

    private void deletePasswordEntryFromKey(JSONArray entries, String key) throws JSONException {
        int deletionIndex = passwordEntryIndex(entries, key);
        if (deletionIndex >= 0)
            entries.remove(deletionIndex);
    }

    private void addPasswordEntry(JSONArray entries, PasswordInfo info) throws JSONException {
        JSONObject json = info.asJsonObject();
        if (json == null) {
            throw new JSONException("Unable to create JSON object from password info");
        }

        entries.put(json);
    }

    /**
     * Closes the password database and makes things secure.
     */
    void lock() {
        rawJson = null;
        activeMasterPassword = null;
        // NOTE: nothing else to do for now.
    }
}

File: Trinity.Runtime/platform_src/ios/elastOS/Classes/PasswordDatabaseInfo.swift
/*
* Copyright (c) 2020 Elastos Foundation
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in all
* copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
* SOFTWARE.
*/

import Foundation

private class PasswordDatabaseInfoApplication {
    private let appID: String
    var passwordEntries = Array<PasswordInfo>()
    
    init(appID: String) {
        self.appID = appID
    }
    
    public static func fromDictionary(_ dict: Dictionary<String, Any>, appID: String) throws -> PasswordDatabaseInfoApplication {
        let app = PasswordDatabaseInfoApplication(appID: appID)
        
        guard let entriesDict = dict["passwordentries"] as? Array<Dictionary<String, Any>> else {
            throw "Invalid dictionary for password database. No passwordentries field"
        }
        
        for entryDict in entriesDict {
            let passwordInfo = try PasswordInfoBuilder.buildFromType(jsonObject: entryDict)
            passwordInfo.appID = appID
            app.passwordEntries.append(passwordInfo)
        }
        
        return app
    }
        
    public func asDictionary() -> Dictionary<String, Any> {
        var dict = Dictionary<String, Any>()
        var passwordEntriesArray = Array<Dictionary<String, Any>>()
        for entry in passwordEntries {
            passwordEntriesArray.append(entry.asDictionary()!)
        }
        dict["passwordentries"] = passwordEntriesArray

        return dict
    }
    
    fileprivate func deletePasswordEntryFromKey(key: String) throws {
        let deletionIndex = try passwordEntryIndex(key: key)
        if deletionIndex >= 0 {
            passwordEntries.remove(at: deletionIndex)
        }
    }

    fileprivate func addPasswordEntry(info: PasswordInfo) {
        passwordEntries.append(info)
    }
    
    fileprivate func passwordEntry(key: String) throws -> PasswordInfo {
        guard let info = passwordEntries.first(where: { info in
            info.key == key
        }) else {
            throw "No password info found for key \(key)"
        }
        
        // READ-ONLY
        info.appID = appID
        
        return info
    }

    fileprivate func passwordEntryIndex(key: String) throws -> Int {
        guard let index = passwordEntries.firstIndex(where: { info in
            info.key == key
        }) else {
            throw "No password info found for key \(key)"
        }
        return index
    }

    fileprivate func keyInPasswordEntries(key: String) -> Bool {
        return ((try? passwordEntryIndex(key: key)) ?? -1) >= 0
    }
}

/**
 * Database JSON format:
 *
 * {
 *     "applications": {
 *          "APPIID1": {
 *              "passwordentries": [
 *                  {
 *                      RAW_USER_OBJECT
 *                  }
 *              ]
 *          }
 *     }
 * }
 *
 * We work directly with raw JSONObjects to make it easier later to maintain the structure, add new fields,
 * handle specific or missing items.
 */
class PasswordDatabaseInfo {
    fileprivate var applications = Dictionary<String, PasswordDatabaseInfoApplication>()
    //var rawJson: Dictionary<String, JSONObject>?
    var activeMasterPassword: String? = nil
    var openingTime: Date

    private init() {
        openingTime = Date()
    }

    static func createEmpty() -> PasswordDatabaseInfo {
        return PasswordDatabaseInfo()
    }

    public static func fromDictionary(_ dict: Dictionary<String, Any>) throws -> PasswordDatabaseInfo {
        let info = PasswordDatabaseInfo()
        
        guard let appsDict = dict["applications"] as? Dictionary<String, Dictionary<String, Any>> else {
            throw "Invalid dictionary for password database. No applications field"
        }
        
        for appDict in appsDict {
            let app = try PasswordDatabaseInfoApplication.fromDictionary(appDict.value, appID: appDict.key)
            info.applications[appDict.key] = app
        }
        
        return info
    }
    
    public func asDictionary() -> Dictionary<String, Any> {
        var dict = Dictionary<String, Any>()
        var applicationsDict = Dictionary<String, Any>()
        for app in applications {
            applicationsDict[app.key] = app.value.asDictionary()
        }
        dict["applications"] = applicationsDict
        
        return dict
    }

    public func getPasswordInfo(appID: String, key: String) throws -> PasswordInfo? {
        if let appIDContent = try getAppIDContent(appID) {
            return (try? appIDContent.passwordEntry(key: key)) ?? nil
        }
        else {
            // No entry for this app ID yet, so we can't find the requested key
            return nil
        }
    }

    public func setPasswordInfo(appID: String, info: PasswordInfo) throws {
        var appIDContent = try getAppIDContent(appID)
        if (appIDContent == nil) {
            // No entry for this app ID yet, create one and add it
            appIDContent = PasswordDatabaseInfoApplication(appID: appID)
            applications[appID] = appIDContent
        }

        if (appIDContent!.keyInPasswordEntries(key: info.key)) {
            // This entry already exists. Delete it first before re-adding its updated version.
            try appIDContent!.deletePasswordEntryFromKey(key: info.key)
        }
        appIDContent!.addPasswordEntry(info: info)
    }

    public func getAllPasswordInfo() throws -> [PasswordInfo] {
        var infos = [PasswordInfo]()
        var it = applications.keys.makeIterator()
        while let appID = it.next() {
            if let appIDContent = try getAppIDContent(appID) {
                for entry in appIDContent.passwordEntries {
                    infos.append(entry)
                }
            }
        }
        return infos
    }

    public func deletePasswordInfo(appID: String, key: String) throws {
        if let appIDContent = try getAppIDContent(appID) {
            try appIDContent.deletePasswordEntryFromKey(key: key)
        }
        else {
            // No entry for this app ID yet, so we can't find the requested key
        }
    }

    private func getAppIDContent(_ appID: String) throws -> PasswordDatabaseInfoApplication? {
        if (applications.keys.contains(appID)) {
            return applications[appID]
        }
        else {
            return nil
        }
    }
    
    /**
     * Closes the password database and makes things secure.
     */
    func lock() {
        applications.removeAll()
        activeMasterPassword = nil
        // NOTE: nothing else to do for now.
    }
}
